src/h3lib/lib/polygon.c

```diff
@@ -93,3 +95,172 @@ bool pointInsidePolygon(const GeoPolygon* geoPolygon, const BBox* bboxes,
+            candidates[containerCount] = polygon;
+            candidateBBoxes[containerCount] = &bboxes[index];
+            containerCount++;
+        }
+        polygon = polygon->next;
+        index++;
+    }
+
+    // Set the initial return value to the first candidate
+    LinkedGeoPolygon* parent = candidates[0];
+
+    // If we have multiple candidates, they must be nested inside each other.
+    // Find the innermost polygon by taking the candidate with the most
+    // containers in the candidate list.
+    if (containerCount > 1) {
+        int max = -1;
+        for (int i = 0; i < containerCount; i++) {
+            int count = countContainers(candidates[i]->first, candidates,
+                                        candidateBBoxes, containerCount);
+            if (count > max) {
+                parent = candidates[i];
+                max = count;
+            }
+        }
+    }
+
+    // Free allocated memory
+    free(candidates);
+    free(candidateBBoxes);
+
+    return parent;
+}
+
+/**
+ * Normalize a LinkedGeoPolygon in-place into a structure following GeoJSON
+ * MultiPolygon rules: Each polygon must have exactly one outer loop, which
+ * must be first in the list, followed by any holes. Holes in this algorithm
+ * are identified by winding order (holes are clockwise), which is guaranteed
+ * by the h3SetToVertexGraph algorithm.
+ *
+ * Input to this function is assumed to be a single polygon including all
+ * loops to normalize. It's assumed that a valid arrangement is possible.
+ *
+ * @param root Root polygon including all loops
+ * @return     0 on success, or an error code > 0 for invalid input
+ */
+int normalizeMultiPolygon(LinkedGeoPolygon* root) {
+    // We assume that the input is a single polygon with loops;
+    // if it has multiple polygons, don't touch it
+    if (root->next) {
+        return NORMALIZATION_ERR_MULTIPLE_POLYGONS;
+    }
+    int loopCount = countLinkedLoops(root);
+    // Early exit if there's only one loop
+    if (loopCount <= 1) {
+        return NORMALIZATION_SUCCESS;
+    }
+    int resultCode = NORMALIZATION_SUCCESS;
+
+    // Create an array to hold all of the inner loops. Note that
+    // this array will never be full, as there will always be fewer
+    // inner loops than outer loops.
+    LinkedGeoLoop** innerLoops = calloc(loopCount, sizeof(LinkedGeoLoop*));
+    assert(innerLoops != NULL);
+
+    // Create an array to hold the bounding boxes for the outer loops
+    BBox* bboxes = calloc(loopCount, sizeof(BBox));
+    assert(bboxes != NULL);
+
+    // Get the first loop and unlink it from root
+    LinkedGeoLoop* loop = root->first;
+    initLinkedPolygon(root);
+
+    // Set up iteration variables
+    LinkedGeoPolygon* polygon = NULL;
+    LinkedGeoLoop* next;
+    int innerCount = 0;
+    int outerCount = 0;
+
+    // Iterate over all loops, moving inner loops into an array and
+    // assigning outer loops to new polygons
+    while (loop) {
```

Comment:
Is there any way to refactor these loops? I find them to be the hardest to grasp conceptually. A good practice I picked up from C++ was to avoid loops when an [std::algorithm](https://en.cppreference.com/w/cpp/algorithm) was available to replace it. That made code flow more idiomatic. This piece of code may be doing something like `std::partition`. Perhaps other parts are more like `std::transform`. If you find a matching pattern, consider refactoring with a clear name, such as `selectInnerLoops`.

Reply:
It's not really either - it's like `partition` but it handles the two groups differently. This loop populates the `innerLoops` array, but it attaches the outer loops to a linked structure instead, and additionally calculates the `bbox` for each outer loop (we don't need it for the inner). I could move this into function, but I'd need to pass in the three different data structures to populate, which make that seem weird.

I'll think about this, but I already broke out several pieces, and I think passing the necessary arguments around is going to make this harder to follow with more functions, not easier.